### ISSUE TITLE
Transaction sync failure on invalid entry date

### ISSUE THREAD

**Issue by joshxg** [openmsupply/mobile#987]
Build Number: 2.3.3

Description:

Currently it is possible that a `Transaction` has no `entryDate` on incoming sync.
When this occurs, parsing the date returns `null`, which is then used to set the field which is not optional, causing an error and failing to integrate this record. This makes initialization impossible.

Reproducible:

Yes

Reproduction Steps:
Create a `transact` record in mSupply with no `entry_date` field (00/00/000).
Try to sync this record

Comments:

Although this should be fixed in mSupply, it exists and needs to be handled.

The default field is to just set the current date, so the easiest fix is to simply set that as the `entryDate` on incoming sync.

@andreievg @Chris-Petty - just double chechking this is the solution we should go with and not others such as ignoring the record?

**Comment by andreievg** [comment]
Can we just make the field optional? I've seen similar with customer requisition, should also set that as optional?

**Comment by joshxg** [comment]
From what I understand a modal can have fields which are either optional or have a default  (exclusive)? I think this would introduce a lot more code changes and possibilities for regressions if anything is missed as well as including migration code?

Also having these records have no value for this field on incoming sync is a bug that should be fixed?

**Comment by Chris-Petty** [comment]
Have you had a chance to investigate where in mSupply Desktop this problem is originating from?

>The default field is to just set the current date, so the easiest fix is to simply set that as the entryDate on incoming sync.

I'd hazard a guess that the default (code written 3 years ago) was trying to handle this situation, so that likely the best way to go. As you said, you'd have to crawl the codebase putting in null checks where needed and also make sure no where was relying on that default (perhaps createRecord for transactions).

**Comment by joshxg** [comment]
> Have you had a chance to investigate where in mSupply Desktop this problem is originating from?
> 

Not yet - I have an issue asking if anyone has any idea where to start

> I'd hazard a guess that the default (code written 3 years ago) was trying to handle this situation, so that likely the best way to go. As you said, you'd have to crawl the codebase putting in null checks where needed and also make sure no where was relying on that default (perhaps createRecord for transactions).

Sure, thanks

**Comment by wlthomson** [comment]
I've added some screenshots of the problematic data to the mSupply issue (https://github.com/sussol/msupply/issues/3401). An mSupply detective might be able to figure it out from there.

**Comment by Chris-Petty** [comment]
> Other date fields, e.g. `confirmDate`, are optional, so I think that seems like a good way to go for consistency. It's also a good idea to enforce things more rigorously on the desktop side, but the mobile app should be resilient against badly formed data. Will go ahead with this and see how it goes.

The fields aren't same in the workflow, we'd expect confirmed date (or finalised date if that's anywhere) to be undefined until a user changes the state. `entryDate` and `createdDate` (they're equivalent 😑) should be set on record creation. Good way to hunt is just to search for `CREATE RECORD([transact])` or `SAVE RECORD([transact])` on mSupply desktop (I'll comment that in the issue there). 

Going against my earlier comment, using default: (currentDate) on the schema just means mobile can't possibly be the source of the problem. Due to this it shouldn't be possible that mobile has caused this problem. We'd lose that safety net in mobile if you changed it to optional.

**Comment by wlthomson** [comment]
OK, my summary taking into account all the above comments.

- problematic date fields can be undefined in the mSupply server datafile, causing problems with sync.
- fixing this with `optional: true` is not suitable, as:
  - these fields should always be defined
  - if these fields are not defined, something has gone wrong, which should be handled not ignored
  - making the fields optional may introduce regressions.
- Fixing only mSupply side is not suitable, as we should not predicate mobile sync on the basis that all the server data is good.

As @joshxg said, the bug here is essentially that certain records have no value for a required field on incoming sync. This is the bug that should be fixed. I think a good solution is to reject these records (this should be logged). IMO using a default value for the problematic fields (with e.g. the current date) is not a good idea as this could cause regressions and other problems.

TLDL; 

Update incoming sync to chuck the bad records out and remove them from the sync out queue alongside other records. The records will stay on the server, but will not be synced to the mobile (until the problematic fields are fixed).

@Chris-Petty @andreievg @joshxg Does this sound good?

**Comment by joshxg** [comment]
If records are ignored by mobile, I assume this means they are not pulled again until the record has been changed? From my little knowledge of sync there is no response to ignoring a record, therefore the server does not know and cannot log anything. 

(btw this change just requires adding the field to the sanity check in incoming sync, I believe)

**Comment by wlthomson** [comment]
> If records are ignored by mobile, I assume this means they are not pulled again until the record has been changed? From my little knowledge of sync there is no response to ignoring a record, therefore the server does not know and cannot log anything.
> 
> (btw this change - ignoring records - just requires adding the field to the sanity check in incoming sync, I believe)

I actually meant log on mobile, not desktop. From what I understand, mobile logging isn't used extensively, but this seems like a case where it is a good idea.

I may be wrong, but I think sanity check may not be appropriate, based on the below snippet:

```
  const hasRequiredFields = requiredFields[recordType].canBeBlank.reduce(
    (containsAllFieldsSoFar, fieldName) =>
      containsAllFieldsSoFar &&
      record[fieldName] !== null && // Key must exist.
      record[fieldName] !== undefined, // Field may be undefined.
    hasAllNonBlankFields
  ); // Initialise |containsAllFieldsSoFar| as result from |hasAllNonBlankFields|.
  return hasRequiredFields;
```

This could be updated a bit to handle dates (i.e. sanity check based on `parseDate(fieldName)==null`), but it would require doing some dynamic type checking which isn't currently done.

**Comment by Chris-Petty** [comment]
@wlthomson Do agree we need good logging, there's a double digit issue for that! #39

> * problematic date fields can be undefined in the mSupply server datafile, causing problems with sync.
> * fixing this with `optional: true` is not suitable, as:
>   
>   * these fields should always be defined
>   * if these fields are not defined, something has gone wrong, which should be handled rather than ignored
>   * making the fields optional may introduce regression

And requires new code, as the current default is what handles setting the date on transaction creation in mobile. May need null checks around the place as well.

> * fixing only mSupply side is not suitable, as we should not predicate mobile sync on the basis that all the server data is good

That's a new design decision IMO. Having required fields block sync has allowed us to presume that the data is valid when implementing much of mobile. If we extended this to other fields, say `TransactionBatch.expiry_date`, we'd have have them silently ignored. The user could then finalise the transaction without that line. Then we're deep in a mess.

Though interestingly apparently that specific case can be both, probs a mistake (a few others there too):
![image](https://user-images.githubusercontent.com/7684221/58606033-0e953e80-82ee-11e9-819c-3bafe6a73bb1.png)

Anyway. For transactions or stocktakes, I think ignoring it is bad. Especially without logging, we need to be well aware of the problem when it's occurring and blocking sync does this well. Not blocking sync allows the user to get confused by their data coming in (or not). This goes doubly for anything that affects ledger like transactionBatches.

**Comment by joshxg** [comment]
It's good that this bug has brought up some good discussions, however we do need a resolution for right now as this does completely break initializing when it occurs and implementing some solid logging feature, acking records synced etc.

From what I can see the three options (for *right now*, not for sync adjustments in the future)  are: 

- Just set a default value (only really current date) for required dates that are invalid on incoming sync, in the incoming sync method (mouthful, sorry)
- Set the field as optional and scan for regressions
- Reject the records

Unless someone else has a good solution to fix this (probably before tomorrow for 2.3.4)

**Comment by wlthomson** [comment]
I've got a PR open. The fix is to reject to records. The idea was to get a short term fix and get some feedback on it for a later better fix. @joshxg have bumped you as an additional reviewer.

**Comment by Chris-Petty** [comment]
My reservations:
1. Allows the root of the problem to be ignored until the customer says "Why are my reports all wrong" if they're doing anything entry date based. Data integrity comes first. Are the problem transactions CIs generated from SIs? Entry date should essentially be the finalise date from the SI. That fixes the data. If that's possible that should be done before any of these other approaches. Has this been investigated?
2. Short term fixes are rarely short term. Lots of short term fixes tend to result it crap code. Are you going to champion chasing up fixing this in the medium term? Keep in mind I've been pushing the datatable rewrite for well over a year.

I'll let this go as a short term fix if the answers are "yes" and "yes".

**Comment by wlthomson** [comment]
`Data integrity comes first.`

I agree. All other solutions I can see involve arbitrarily editing the missing field to sanitise it, which actually papers over the problem rather that revealing it.

`"Why are my reports all wrong" if they're doing anything entry date based.`

As I see it that's an mSupply problem, not mobile. Mobile is concerned with enforcing integrity and its own constraints on its data. If mSupply gives it misformed data, mobile should log it, but imo it's not the frontend's responsibility to handle a problem coming from the API.

`Entry date should essentially be the finalise date from the SI. That fixes the data.`

The transactions were a mix of types (see the mSupply issue linked in my first comment), so at this stage I don't think there is an obvious cause. 

`Are you going to champion chasing up fixing this in the medium term?`

I'm happy with that, with the caveat that I'm not sure where else we can go re this except for log and perhaps integrate with the hub etc.  Imo it's acceptable for each sender to ensure the data it sends is correctly formatted and in turn each receiver to reject data that is malformed. Even if mSupply sends malformed data, I don't think we should get into the business of coupling mobile and mSupply to the extent that they are "fixing" each other's data. I think that will lead us down a bad path where we actually increase the difficulty of debugging and tracking down data corruptions. 

TLDR; I think mobile should log to indicate a problem, but that it is the responsibility of mSupply to fix the cause.

**Comment by Chris-Petty** [comment]
> I agree. All other solutions I can see involve arbitrarily editing the missing field to sanitise it, which actually papers over the problem rather that revealing it.

The compromise before this issue was that mobile assumes that records are valid for required fields such as date, nameID, ID. If not, block sync. This means we get the handbrake on and we fix the source of the problem. That's rough and blunt, but it has saved us from writing massive amounts of validation for various edge cases, the safe handling, UI to communicate it... Your solution literally papers over it until data integrity rears its head later on in reporting or "missing" transaction support tickets.

> Mobile is concerned with enforcing integrity and its own constraints on its data. If mSupply gives it misformed data, mobile should log it, but imo it's not the frontend's responsibility to handle a problem coming from the API.

But mobile isn't simply a front end to a server's API doing the logic. That's the point of the synchronisation system. With the proposed change mobile is going to have orphan requisitionItems. These will make a placeholder for the requisition that has been silently ignored. Fortunately the placeholder won't show. But for supplier invoices this will just make it look like they're missing incoming orders. Then they can't/shouldn't work with mobile and we're a step obfuscated from helping them. They might do a stocktake to correct stock levels, then if we fix the supplier invoice they get double entry. Customer invoice probably worse, risking re-entry outright. If we were to do ledger checks in mobile it'll show problems, as there will be placeholder transactions that may have been finalised in the past but now have no status.

TL;DR: If you want the users to be safe to continue their work flow, I think the safer solution is to let it sync without entry date being required (i.e. canBeBlank, optional field) with care to make sure it's set appropriately (probably just in create record?). Shouldn't be much more work and doesn't invite reentry. If you truly want purity of mobile not handling server data/api problems, find and fix the source in server (and/or mobile if that's the case) and work out a reasonable way to address the live data on the server. Which should be done either way.

**Comment by wlthomson** [comment]
I've closed the PR as requested by @Chris-Petty.

I've had a chat with @Chris-Petty and we've agreed that the best approach requires:

- not mutating the malformed data
  - no obvious way to choose a placeholder date
  - mobile shouldn't mutate and sync back "untouched" data
- not rejecting the malformed data
  - see https://github.com/openmsupply/mobile/issues/987#issuecomment-498924821

The simplest solution seems to be to update realm schema for these fields to be optional, and to set them in `createRecord` to `null`. This might also require updating some other logic to prevent regressions.

**Comment by joshxg** [comment]
@wlthomson Have moved this back to in progress, still ok with it assigned to you?

**Comment by wlthomson** [comment]
Sounds good. Have put it back into my backlog.

**Comment by wlthomson** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [Mobile sync bug fix](https://trello.com/c/9ZYBEDLd/122-mobile-sync-bug-fix)

**Comment by wlthomson** [comment]
Not sure if people aware, but this looks like the likely culprit: https://github.com/sussol/msupply/issues/3422.

**Comment by joshxg** [comment]
Closed by #1047

**Comment by joshxg** [comment]
## Testing

Steps to reproduce or otherwise test the changes of this PR:

- [x] Incoming sync handles default transaction entry date (`00-00-0000`).
- [x] Outgoing sync handles `null` transaction entry dates.
- [x] Supplier invoice page correctly displays transactions with `null` entry dates.
- [x] Supplier invoice page correctly displays transactions with `null` entry dates.
- [x] Customer invoice page correctly displays transactions with `null` entry dates.
- [x] Customer invoice page correctly displays transactions with `null` entry dates.
